[doc] Add guide to setup and run ROM E2E tests (#17483)
lowRISC/opentitan: doc/guides/getting_started/src/e2e_regression_setup.md

> # ROM E2E Regressions setup

Reviewer: It would also be useful to have a link pointing to the ROM E2E test plan. Not many people are familiar with how things are setup in the repo.

Author: done

> # ROM E2E Regressions setup

Reviewer: Maintainers and downstream consumers are also interested in understanding how the test setup works so that they can add or modify existing test cases. It would be useful to have a diagram showing all the components (e.g. DUT, opentitanlib, Bazel, etc).

Author: Done

This guide will help you to setup an environment to run the ROM E2E tests in an Ubuntu 20.04 machine.

## Hardware setup
### FPGA
The E2E ROM tests run in the CW310 FPGA board with the following configuration:
 - The power supply connected to the J11 connector.

>  - An USB cable connecting the Host PC to the USB-C Data (J8) connector. 

Reviewer 2: ```suggestion
 - A USB cable connecting the Host PC to the USB-C Data (J8) connector. 
```

 - The USRDIP#0 (SW4 header) is set to 0.
 - The rest of the USRDIPs are set to 1.

>  - The S3 switch (next to USB connector J9) is set to USB

Reviewer 2: ```suggestion
 - The S3 switch (next to USB connector J9) is set to USB.
```

>  - The S2 switch (next to JTAG header) is set to +3.3V

Reviewer 2: ```suggestion
 - The S2 switch (next to JTAG header) is set to +3.3V.
```

>  - Optionally a USB cable connecting the host PC to the USB-C J9 connector for future tests that will test the USB block.

Reviewer 2: ```suggestion
 - Optionally, a USB cable connecting the host PC to the USB-C J9 connector for future tests that will test the USB block.
```

  
### JTAG adapter

> Some tests will use the JTAG interface to interact with Opentitan, as the CW310 doesn't have JTAG adapter integrated for Opentitan (FTDI) we can use an external adapter (Olimex ARM-USB-TINY-H JTAG) connected to the J13 header in the CW310.

Reviewer 2: ```suggestion
Some tests will use the JTAG interface to interact with Opentitan. Since the CW310 doesn't have a USB to JTAG adapter we can use an external adapter (Olimex ARM-USB-TINY-H JTAG) connected to the J13 header in the CW310.
```

### Full hardware setup
![Full hardware setup](CW310_setup.png)

## Software setup 
The software dependencies are covered by other guides in this book and are linked below.
- Clone the Opentitan repository and install the software dependencies: [Get started](README.md).
- Setup the FPGA: [FPGA guide](./setup_fpga.md).
- Install Vivado to be able to splice bitstreams: [Installing Vivado](install_vivado/README.md).

**Note**: Make sure that you added the udev rules for the [FPGA](install_vivado/README.md#device-permissions-udev-rules) board and the [JTAG](setup_fpga.md#device-permissions-udev-rules) adapter.

> ## Running the ROM E2E tests

Reviewer: nit: Since not everyone is familiar with Bazel. I think it would be useful to have a reference to the build file containing the test targets, i.e. `sw/device/silicon_creator/rom/e2e/BUILD`

> To run the ROM E2E tests you just need to run the command `bazel test` pointing at the `sw/device/silicon_creator/rom/e2e` directory.

Reviewer 2: ```suggestion
To run the ROM E2E tests you just need to run the command `bazel test`, with a test target for the package `//sw/device/silicon_creator/rom/e2e`.
```

Reviewer 2: @cfrantz I'm not quite sure how to phrase that.

Reviewer 3: """
To run the full suite of ROM E2E tests, execute `bazel test` with the label wildcard `//sw/device/silicon_creator/rom/e2e/...`.
"""

Author: Done

```console
cd ${REPO_TOP}
./bazelisk.sh test --define DISABLE_VERILATOR_BUILD=true --define bitstream=vivado --test_tag_filters=-verilator,-dv,-broken --build_tests_only //sw/device/silicon_creator/rom/e2e/...
```

> This bazel command will build a bitstream, splice it with the different OPT configurations, load the FPGA and run all the ROM_E2E tests available under `sw/device/silicon_creator/rom/e2e`.

Reviewer: s/OPT/OTP/

> If you want to save time you can use cached bitstream by changing the `bitstream=vivado` to `bitstream=gcp_splice`.

Reviewer: Does using `bitstream=gcp_splice` take the ROM version associated with the repo `HEAD`? If not, it is work mentioning here in an additional note.

Author: I believe so, can you confirm it @alphan ?

# ROM E2E Regressions setup
This guide will help you to setup an environment to run the ROM E2E tests in an Ubuntu 20.04 machine.

## Introduction

> The [ROM](https://opentitan.org/book/doc/security/specs/secure_boot/index.html?highlight=ROM#rom-rom) is first boot stage of secure boot and it is placed in the ROM memory which means that it cannot be updated after manufacturing at all.

Reviewer 4: ```suggestion
The [ROM](https://opentitan.org/book/doc/security/specs/secure_boot/index.html?highlight=ROM#rom-rom) is the first boot stage of the secure boot flow, and by nature, it cannot be updated after manufacturing.
```

ROM = Read Only Memory :)

> The ROM E2E (End-to-End) tests validates the ROM features and can be used for regression tests.

Reviewer 4: ```suggestion
The ROM E2E (End-to-End) tests validate the ROM features and can be used for regression tests.
```

Author: Done

Each test is divided into components:

>  - **Opentitantool**, it is a Rust program that provide a common interface to interact with Opentitan. Opentitantool uses opentitanlib which provides the abstractions to communicate with Opentitan.

Reviewer 4: ```suggestion
 - **`opentitantool`**, a Rust program that provides a common interface to interact with an OpenTitan device. `opentitantool` uses `opentitanlib` which provides abstractions to communicate with an OpenTitan device.
```

Author: Done

>  - **Host test** software, it is a Rust program built by bazel that actively interacts with the device software via UART, SPI or JTAG to perform the test and check whether it passed or not. To avoid code duplication the tests uses the `opentitanlib` to communicate with opentitan.

Reviewer 4: ```suggestion
 - **Host test** software, any Rust program (built by `bazel` to run on the host platform, e.g. an x86 binary, and linked with `opentitanlib`) that actively interacts with an OpenTitan device over any HW interface (e.g., UART, SPI or JTAG, etc.) to perform the test functions and check their results.```

Author: Done

>  - **Device test** software, it is a C program built by bazel that run in the flash after the ROM and perform checking in the HW and/or memories to check the configurations performed by the ROM.

Reviewer 4: ```suggestion
 - **Device test** software, a C program (built by `bazel` to run on an OpenTitan device) that is loaded into flash, and booted by the ROM stage. It checks the HW and/or memory configurations performed by the ROM and relays information to the Host test software.
```
The tests are loaded into flash but are copied into SRAM before they are run by the C runtime, so technically they run out of SRAM, but are loaded into flash.

Author: Interesting, out of curiosity, why we run them from SRAM, performance requirements?

>  - **FPGA bitstream**, it is built by bazel which using vivado.

Reviewer 4: ```suggestion
 - **FPGA bitstream**, OpenTitan HW implementation, built by `bazel` and `vivado`.
```

Author: Done.

 - **ROM Image**, it is a C program built by bazel which is the DUT (Device Under Test).

>  - **OTP image**, contains secure configurations read by the ROM to define workflows such as the lifecycle stage, so for some tests may be executed with several different OTP images to test all corner cases.

Reviewer 4: ```suggestion
 - **OTP image**, contains HW and SW configurations, some of which control execution paths in the ROM, such as, the lifecycle stage. As a result, we run some tests with several different OTP images to increase coverage of ROM execution paths.
```
maybe would be good to [link](https://opentitan.org/book/hw/ip/otp_ctrl/index.html#direct-access-memory-map) to the OTP `CREATOR_SW_CFG` table here?

Author: Done

![Architecture](ROM_E2E_Tests.svg)

All the existing tests are documented in the [ROM E2E Testplan](https://github.com/lowRISC/opentitan/blob/master/sw/device/silicon_creator/rom/data/rom_e2e_testplan.hjson).
## Hardware setup
### FPGA

> The E2E ROM tests run in the CW310 FPGA board with the following configuration:

Reviewer 4: ```suggestion
The E2E ROM tests run on the CW310 FPGA board with the following configuration:
```

Author: Done

 - The power supply connected to the J11 connector.

>  - A USB cable connecting the Host PC to the USB-C Data (J8) connector which is connected to the SAM3U in the board.

Reviewer 4: ```suggestion
 - A USB cable connecting the Host PC to the USB-C Data (J8) connector which is connected to the SAM3U microcontroller on the CW310 FPGA board.
```

Author: Done

 - The USRDIP#0 (SW4 header) is set to 0.
 - The rest of the USRDIPs are set to 1.
 - The S3 switch (next to USB connector J9) is set to USB.
 - The S2 switch (next to JTAG header) is set to +3.3V.

>  - Optionally for future tests, a USB cable connecting the host PC to the USB-C J9 connector, which is connected to the Opentitan `usbdev` IP block.

Reviewer 4: ```suggestion
 - Optionally for future tests, a USB cable connecting the host PC to the USB-C J9 connector, which is connected to the OpenTitan `usbdev` IP block.
```

Author: Done

### JTAG adapter

> Some tests will use the JTAG interface to interact with Opentitan, as the CW310 doesn't have a embedded JTAG adapter, we can use an external adapter (Olimex ARM-USB-TINY-H JTAG) connected to the J13 header in the CW310.

Reviewer 4: ```suggestion
Some tests will use the JTAG interface to interact with OpenTitan, as the CW310 doesn't have a embedded JTAG adapter, we can use an external adapter (Olimex ARM-USB-TINY-H JTAG) connected to the J13 header in the CW310.
```

Author: Done

  
![Architecture](ROM_E2E_Tests.svg)

All the existing tests are documented in the [ROM E2E Testplan](https://github.com/lowRISC/opentitan/blob/master/sw/device/silicon_creator/rom/data/rom_e2e_testplan.hjson).
## Hardware setup
### FPGA
The E2E ROM tests run in the CW310 FPGA board with the following configuration:
 - The power supply connected to the J11 connector.
 - A USB cable connecting the Host PC to the USB-C Data (J8) connector. 
 - The USRDIP#0 (SW4 header) is set to 0.
 - The rest of the USRDIPs are set to 1.
 - The S3 switch (next to USB connector J9) is set to USB.
 - The S2 switch (next to JTAG header) is set to +3.3V.
 - Optionally, a USB cable connecting the host PC to the USB-C J9 connector for future tests that will test the USB block.
  
### JTAG adapter
Some tests will use the JTAG interface to interact with Opentitan, as the CW310 doesn't have a embedded JTAG adapter, we can use an external adapter (Olimex ARM-USB-TINY-H JTAG) connected to the J13 header in the CW310.

### Full hardware setup

> ![Full hardware setup](CW310_setup.png)

Reviewer 3: Just a point of clarification on the cabling shown in the image:
- The red USB cable connects to the SAM3U control chip on the board.
- The white USB cable connects to the user USB port and this port connects to the OT `usbdev` IP.

AFAIK, we currently do not have any tests that use the `usbdev` IP.  I believe such tests are under development at lowRISC.

## Software setup 
The software dependencies are covered by other guides in this book and are linked below.
- Clone the Opentitan repository and install the software dependencies: [Get started](README.md).
- Setup the FPGA: [FPGA guide](./setup_fpga.md).
- Install Vivado to be able to splice bitstreams: [Installing Vivado](install_vivado/README.md).

**Note**: Make sure that you added the udev rules for the [FPGA](install_vivado/README.md#device-permissions-udev-rules) board and the [JTAG](setup_fpga.md#device-permissions-udev-rules) adapter.

## Running the ROM E2E tests
The ROM E2E tests are listed in the bazel [BUILD file](https://github.com/lowRISC/opentitan/blob/master/sw/device/silicon_creator/rom/e2e/BUILD), alternatively they can be listed with the command:
```sh
bazel query 'attr(tags, cw310, tests(//sw/device/silicon_creator/rom/e2e/...))
``` 
Once the hardware setup is done and the FPGA is connected to the USB, the ROM E2E tests can be executed with the command `bazel test` pointing at the `sw/device/silicon_creator/rom/e2e` directory.
```sh
cd ${REPO_TOP}

> bazel test --define DISABLE_VERILATOR_BUILD=true --define bitstream=vivado --test_tag_filters=-verilator,-dv,-broken --build_tests_only //sw/device/silicon_creator/rom/e2e/...

Reviewer 3: I'm a bit loathe to tell people to use `bitstream=vivado`.  That causes bazel to invoke vivado and synthesize the bitstream.  If you're modifying the RTL, then this is the right thing to do.  However, if you're only modifying the software, then `bitstream=gcp_splice` is more likely what you wwant.

Author: I changed the text to make it clear that the `gcp_splice` is the desirable way, but `vivado` is also possible.
